docs/static-evaluation.md

```diff
@@ -0,0 +1,388 @@
+
+### Provider Iteration
+
+In [#300](https://github.com/opentofu/opentofu/issues/300), users describe how supporting for_each for configuring providers will allow much DRYer and simpler configurations.
+
+> [!Note]
+> Please read [Provider References](provider-references.md) before diving into this section!
+
+#### Proposed Changes
+
+The first change is to support static variables in the provider config block.  This can then be extended to support for_each/count and be expanded at the end of the `config.NewModule()` function, similar to how module.ProviderLocalNames is generated.  This piece is fairly straightforward and can be done relatively easily.
+
+The next step is to support provider aliases indexed by an expression, which is quite a bit trickier.
+Example:
+```hcl
+locals {
+  regions = {"us": "us-east-1", "eu": "eu-west-1"}
+}
+
+
+provider "aws" {
+  for_each = local.regions
+  alias = each.key
+  region = each.value
+}
+
+resource "aws_s3_bucket" "primary" {
+  for_each = local.regions
+  provider = provider.aws[each.key]
+}
+
+module "mod" {
+  source = "./mod"
+  providers {
+    aws = provider.aws[each.key]
+  }
+}
+```
+
+As you can see, the `provider.name[alias]` form is introduced in that example.  This allows providers named "local" or other conflicting names, and clearly shows that it's referencing a particular instance of a given type.
+
+At this point, we don't have a clear path to implementation, but we can enumerate some of the challenges that are faced:
+
+* Introducing an alternate provider address method and updating documentation
+* Provider mappings for modules and resources are hard-coded at config load time on an *unexpanded* view of the config/graph structures
+* Provider configurations are [pruned during graph processing](./provider-references.md#Provider-Workflow)
+
+Potential implementation paths are explored in [Static Evaluation of Providers](./static-evaluation-providers.md)
+
+#### Questions
+
+Should variables be allowed in required_providers now or in the future?  Could help with versioning / swapping out for testing?
+
+In #300, there's also a discussion on allowing variable names in provider aliases.
+Example:
+```
+# Why would you want to do this?  It looks like terraform deprecated this some time after 0.11.
+provider "aws" {
+  alias = var.foo
+}
+```
+
+### Backend Configuration
+
+Once the core is implemented, this is probably the easiest solution to implement.
+
+Notes from initial prototyping:
+* The configs.Backend saves the config body during the config load and does not evaluate it
+* backendInitFromConfig() in command/meta_backend.go is what evaluates the body
+  - This happens before the graph is constructed / evaluated and can be considered an extension of the config loading stage.
+* We can stash a copy of the StaticContext in the configs.Backend and use it in backendInitFromConfig() to provide do the decoding.
```

Comment:
```suggestion
* We can stash a copy of the StaticContext in the configs.Backend and use it in backendInitFromConfig() do the decoding.
```

Reply:
Also, in "to do the decoding", you mean to evaluate the references instead, right?